clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -633,6 +641,9 @@ private void handleListOffsetResponse(TopicPartition topicPartition,
             completedFetch.metricAggregator.record(tp, bytes, recordsCount);
         }
 
+        if (!emptyResponse)
+            subscriptions.movePartitionToEnd(tp);
```

Comment:
When reading messages from different partitions, KStream prefers to read from partitions with smaller timestamps first and only advances the KStream timestamp when it sees at least one message from every partition. So, when a consumer lags (e.g., rewinding), it a partition starves, it will delay the advancing of the timestamp in KStream. So, doing shuffling based on the returned size is a bit better than random shuffling since it's more predictable.